### PR TITLE
Add LayerNorm (1D only)

### LINKED ISSUE #1959: [Feature Request] Layer Normalization
See https://github.com/pytorch/pytorch/issues/1601 for previous discussion on [layer normalization](https://arxiv.org/abs/1607.06450).

### PR DESCRIPTION
**Pull request by Kaixhin**
Closes https://github.com/pytorch/pytorch/issues/1959. Probably needs a good review.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -556,6 +556,22 @@ def batch_norm(input, running_mean, running_var, weight=None, bias=None,
     return f(input, weight, bias)
 
 
+def layer_norm(input, weight=None, bias=None, eps=1e-5):
```
Comment (alykhantejani) [inline torch/nn/functional.py]: we should also move the dim checks here, as this can be called outside of the module.

Comment (alykhantejani) [inline torch/nn/functional.py]: ah, well I guess this becomes tricky with `LayerNormxD`

Comment (alykhantejani) [inline torch/nn/functional.py]: Maybe just this one should go in the functional interface:
```Python
if weight is not None and input.size(1) != self.weight.nelement():
           raise ValueError('got {}-feature tensor, expected {}'
                            .format(input.size(1), self.num_features))
```

File: torch/nn/functional.py [hunk 2]
```diff
@@ -563,6 +563,30 @@ def batch_norm(input, running_mean, running_var, weight=None, bias=None,
...
+    mean = input.mean(1, keepdim=True)
+    std = input.std(1, keepdim=True)
+    output = (input - mean.expand_as(input)) / (std.expand_as(input) + eps)
+
+    # Resize weights and biases to match dims
+    if weight is not None:
+        if input.size(1) != weight.nelement():
+            raise ValueError('got {}-feature tensor, expected {}'
+                             .format(input.size(1), weight.nelement()))
+        else:
```
Comment (alykhantejani) [inline torch/nn/functional.py]: nit: can probable drop the else

**Comment by alykhantejani** [comment]
Thanks Kai, code looks good! Can you also add some entries to `new_module_tests` as done here: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3083 

Could also add checks that the functional `layer_norm` raises an error with size mismatches as done here https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L2287

**Comment by alykhantejani** [comment]
Thanks Kai, code looks good! Can you also add some entries to `new_module_tests` as done here: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3083 

Could also add checks that the functional `layer_norm` raises an error with size mismatches as done here https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L2287

**Comment by Kaixhin** [comment]
Making me feel nostalgic Aly :) Will try to get on with the tests section this weekend.

**Comment by Kaixhin** [comment]
Almost ready - just want to check something first and will comment here again once I'm done.

**Comment by Kaixhin** [comment]
I'm probably being stupid here, but pinging @DmitryUlyanov to explain the difference between instance normalisation and layer normalisation (if indeed there is any).

**Comment by DmitryUlyanov** [comment]
Hi @Kaixhin, to be true, from LayerNorm paper I did not manage to understand how exactly they normalized data with spatial dimensions. So as you say, either Instance norm and layer norm are identical, or they are not and layer norm normalizes each pixel independently as if it was an output of fc layer. 

We tried to replace BatchNorm with InstanceNorm for ImageNet classification and obtained similar accuracy drop to what they report. So this could be an indicator for the first case, but I've never tried the second one so cannot be sure.

**Comment by Kaixhin** [comment]
@DmitryUlyanov Thanks for the extra info. I found [Normalizing the Normalizers](https://arxiv.org/abs/1611.04520), which seems to back up the interpretation that the entire layer is normalised over in layer norm/as opposed to instance norm. Whilst I am using the text for `InstanceNorm{1,2,3}d` for `LayerNorm`, I am adjusting the text for `InstanceNorm{1,2,3}d` to have:

> The mean and standard-deviation are calculated per-dimension separately for each object in a mini-batch. Gamma and beta are learnable parameter vectors of size C (where C is the input size). This can be seen as an extension of layer normalization where statistics are only calculated over `num_features` and NOT all non-batch dimensions.

Is this both descriptive and fair? I would invite you to have a look at the new PR that I've opened to see the exact changes.

**Comment by Kaixhin** [comment]
@alykhantejani Following the discussion in https://github.com/pytorch/pytorch/pull/2112, it seems like it'll take a while to decide what the best thing to do regarding InstanceNorm is, so this PR is now just the 1D version of LayerNorm with a simple implementation, and not touching InstanceNorm.

**Comment by Kaixhin** [comment]
Before we get this in, question about the API (@jekbradbury may want to weight in too) - layer norm doesn't need to have a known input size, since it doesn't require learnable parameters. But for the sake of consistency with most other things, should we have the layer take in a size as input, and an `affine` argument to disable learnable parameters? The functional interface can still take an arbitrary size (and do the checks to make sure if weights/biases are passed that they are of the correct size).

**Comment by jekbradbury** [comment]
I'm personally fine with the slightly magical API here, because I'd only want to have to provide a feature dimension (or use the Module version at all) if I'm using affine parameters.

**Comment by Kaixhin** [comment]
OK @alykhantejani this is self-contained and ready to go - should just need a final review. People can discuss LayerNorm vs. InstanceNorm in the other PR.

File: torch/nn/modules/instancenorm.py [hunk 3]
```diff
@@ -36,12 +39,81 @@ def eval(self):
...
+        self.num_features = num_features
+        self.affine = num_features is not None
+        self.eps = eps
+        if self.affine:
+            self.weight = Parameter(torch.ones(num_features))
+            self.bias = Parameter(torch.zeros(num_features))
+        else:
+            self.register_parameter('weight', None)
+            self.register_parameter('bias', None)
+        self.reset_parameters()
```
Comment (alykhantejani) [inline torch/nn/modules/instancenorm.py]: Not sure if this call is needed as they are already initialized this way

File: torch/nn/modules/instancenorm.py [hunk 4]
```diff
@@ -36,12 +39,81 @@ def eval(self):
...
+                    .format(name=self.__class__.__name__, **self.__dict__))
+        else:
+            return ('{name}(eps={eps})'
+                    .format(name=self.__class__.__name__, **self.__dict__))
+
+    def _check_input_dim(self, input):
+        if input.dim() != 2:
+            raise ValueError('expected 2D input (got {}D input)'
+                             .format(input.dim()))
+        super(LayerNorm, self)._check_input_dim(input)
```
Comment (alykhantejani) [inline torch/nn/modules/instancenorm.py]: is this call needed, there's no implementation of `_check_input_dim` on `Module`

File: torch/nn/modules/instancenorm.py [hunk 5]
```diff
@@ -36,12 +39,81 @@ def eval(self):
...
+
+    def __repr__(self):
+        if self.affine:
+            return ('{name}({num_features}, eps={eps})'
+                    .format(name=self.__class__.__name__, **self.__dict__))
+        else:
+            return ('{name}(eps={eps})'
+                    .format(name=self.__class__.__name__, **self.__dict__))
+
+    def _check_input_dim(self, input):
```
Comment (alykhantejani) [inline torch/nn/modules/instancenorm.py]: where is this called? Perhaps this logic should be moved into the `layer_norm` function?

**Review by alykhantejani — CHANGES_REQUESTED** [review]

File: torch/nn/modules/instancenorm.py (hunk 3 above)
Comment (Kaixhin) [reply torch/nn/modules/instancenorm.py]: Good point - going to just create them as Tensors and then keep this in to a) separate concerns of weight initialisation into one place b) be consistent with `batchnorm.py`.

File: torch/nn/modules/instancenorm.py (hunk 4 above)
Comment (Kaixhin) [reply torch/nn/modules/instancenorm.py]: Ah I see these functions have been removed from BatchNorm and InstanceNorm from when I copied them. It's not needed here, correct.

File: torch/nn/modules/instancenorm.py (hunk 5 above)
Comment (Kaixhin) [reply torch/nn/modules/instancenorm.py]: Yep, will do.

File: torch/nn/functional.py [hunk 6]
```diff
@@ -731,6 +731,29 @@ def batch_norm(input, running_mean, running_var, weight=None, bias=None,
...
+        raise ValueError("Expected 2D tensor as input, got {}D tensor instead.".format(input.dim()))
+
+    mean = input.mean(1, keepdim=True)
+    std = input.std(1, keepdim=True)
+    output = (input - mean) / (std + eps)
+
+    # Resize weights and biases to match dims
+    if weight is not None:
+        if input.size(1) != weight.nelement():
+            raise RuntimeError('got {}-feature tensor, expected {}'
```
Comment (alykhantejani) [inline torch/nn/functional.py]: nit: `got {}D feature tensor` might read better

**Review by alykhantejani — APPROVED** [review]
LGTM - thanks @Kaixhin

**Comment by mys007** [comment]
I got bitten by https://github.com/pytorch/pytorch/issues/2534 here: when the standard deviation happens to be 0, the gradient will become NaN. The following workaround fixes the computation here for me:
```
std = ((input - mean).pow(2).sum(1, keepdim=True).div(input.size(1)-1) + eps).sqrt()
output = (input - mean) / std

**Comment by Kaixhin** [comment]
@mys007 This looks like a wide-reaching problem - would it not affect any other functions which make use of standard deviations (batch norm, instance norm, etc.)? If so then we probably need to have this solution more widely available rather than just specific to layer norm.

**Comment by mys007** [comment]
@Kaixhin It is a wide-reaching issue but IMHO it doesn't affect batchnorm/instancenorm/... as they have their C implementation in the backend, which does not rely on autograd's `sqrt`.

**Comment by Kaixhin** [comment]
@mys007 sorry, only saw first half of your comment before adding this. Are there plans for `std()` to be fixed then? If not, maybe you should submit a PR.

**Comment by mys007** [comment]
@Kaixhin Cool. Just a detail: to be equivalent to Batch/InstanceNorm implementation (https://github.com/zdevito/ATen/blob/master/src/THNN/generic/BatchNormalization.c), one should use biased stddev computation. But even with that, `LN(t)` and `IN(t.unsqueeze(1)).squeeze(1)` for `M=LayerNorm()` and `IN=nn.InstanceNorm1d(1)` have some subtle differences, I assume they are numerical issues.

I haven't found any plans for `std()` in issues. IMHO it should be also done with subgradient, instead of just adding eps. I'll try to look at it later.

**Comment by Kaixhin** [comment]
@mys007 thanks for the input - we'll wait for your findings. There is a related PR (https://github.com/pytorch/pytorch/pull/2112) on LayerNorm vs. InstanceNorm that would be relevant (if you haven't come across it already).

File: torch/nn/functional.py [hunk 7]
```diff
@@ -1178,6 +1178,30 @@ def local_response_norm(input, size, alpha=1e-4, beta=0.75, k=1):
     return input / div
 
 
+def layer_norm(input, weight=None, bias=None, eps=1e-5):
+    if input is not None and input.dim() != 2:
+        raise ValueError("Expected 2D tensor as input, got {}D tensor instead.".format(input.dim()))
+
+    mean = input.mean(1, keepdim=True)
+    # Prevent NaN gradients when sample std is 0 by using alternative standard deviation calculation
+    std = ((input - mean).pow(2).sum(1, keepdim=True).div(input.size(1) - 1) + eps).sqrt()
```
Comment (apaszke) [inline torch/nn/functional.py]: How does this help?

Comment (Kaixhin) [reply torch/nn/functional.py]: @mys007 can you explain this one?

Comment (mys007) [reply torch/nn/functional.py]: Kaixhin's commit was the reaction to my report here: https://github.com/pytorch/pytorch/pull/2019#issuecomment-342787832 . But maybe the `std()` implementation was fixed in the meantime?

**Comment by Kaixhin** [comment]
Closing in favour of https://github.com/pytorch/pytorch/pull/4922
